File: true.c
#include "true.h"
#include "meme.h"
#include "node.h"
#include "printer.h"
#include <assert.h>

static Type const true_type;

static void true_print(Node *node, Printer *p) {
    assert(node->type == &true_type);
    print_atom(p, "#t");
}

static Type const true_type = {
    .name = "true",
    .print = true_print,
    .eval = node_eval_self,
};

Node true_node_storage = {
    .refs = 1,
    .type = &true_type,
};

Node *true_node = &true_node_storage;

File: true.go
package meme

type trueType struct{}

func (me trueType) Apply([]Node, Env) Node {
	panic("true is not an operator")
}

func (me trueType) Eval(Env) Node {
	return me
}

func (trueType) Expand(Env) Parseable {
	return True
}

func (trueType) Print(p *Print) {
	p.Atom("#t")
}

var True trueType
